[DeviceMesh] Isolate pg creation logic in Device Mesh into a separate func `_init_one_process_group`

fduwjj opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #166614

To makes pg cache change easier and code modularization, we isolate the logic of process group creation into a separate function named `_init_one_process_group`.


cc @H-Huang @awgu @wanchaol @fegin @wz337 @wconstab @d4l3k @pragupta @msaroufim @dcci

Differential Revision: [D85920931](https://our.internmc.facebook.com/intern/diff/D85920931)

fduwjj committed:
[DeviceMesh] Isolate pg creation logic in Device Mesh into a separate func _init_one_process_group

[ghstack-poisoned]

lw commented on torch/distributed/device_mesh.py:
```diff
@@ -349,6 +349,81 @@ def _setup_world_group_and_device(self):
+            backend_override: BackendConfig,
+        ) -> Optional[str]:
+            # swap the current dim to the last dim
+            # then reshape to flatten out other dims
```
Nit: I think this comment is outdated, and it was probably me who forgot to remove it when I changed this code.

lw commented on torch/distributed/device_mesh.py:
```diff
@@ -384,86 +459,12 @@ def _init_process_groups(
-                                )
-                            dim_group_names.append(dim_group.group_name)  # type: ignore[union-attr]
+                    if pg_name is not None:
+                        dim_group_names.append(pg_name)
```
I always find this a bit confusing that the pg_name could be None.

If I understand correctly this happens when we create (or when we unflatten?) a mesh that we don't belong to. I think I see why we need to support it, but I worry this could become a pitfall.

To check my understanding, do we agree that (in this code here) either _all_ dims have pg_name == None or _none_ of them does?

If this is the case, could we make it clearer? For example, we could structure the code as follows:
```py
for dim in range(len(layout)):
    ...
    dim_group_names.append(pg_name)  # unconditionally
if any(n is None for n in dim_group_names):
    assert all(n is None for n in dim_group_names)
    return []
return dim_group_names
```
Maybe we could even return None instead of `[]` to further reduce the risk of pitfalls?

lw reviewed (changes requested):
Thanks!

The only blocking comment is the one about moving the `layout.numel() == world_size` check into the new method.

fduwjj committed:
Update on "[DeviceMesh] Isolate pg creation logic in Device Mesh into a separate func `_init_one_process_group`"


To makes pg cache change easier and code modularization, we isolate the logic of process group creation into a separate function named `_init_one_process_group`.


cc H-Huang awgu wanchaol fegin wz337 wconstab d4l3k pragupta msaroufim dcci

[ghstack-poisoned]

lw commented on torch/distributed/device_mesh.py:
```diff
@@ -380,90 +391,80 @@ def _init_process_groups(
+            # numbers of API calls are equal to the number of subgroups for each mesh dimension. In a 2 * 4
+            # mesh, we need to make two API calls per ranks to create all the subgroups.
+            if (
+                getattr(default_group, "bound_device_id", None) is not None
```
Thanks for simplifying this!!

lw reviewed (approved):
Great! I like this a lot!

fduwjj commented on torch/distributed/device_mesh.py:
```diff
@@ -380,90 +391,80 @@ def _init_process_groups(
+            for dim in range(len(layout)):
+                dim_name = mesh_dim_names[dim] if mesh_dim_names else f"dim_{dim}"
+                dim_group_names.append(
+                    DeviceMesh._init_one_process_group(  # type: ignore[arg-type]
```
it complained we are appending None or str while we define dim_group_names to be a list of str.. Let me see if I can fix it.

fduwjj commented:
@fduwjj has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D85920931).
